openlibrary/core/models.py

```diff
@@ -401,10 +428,33 @@ def from_isbn(cls, isbn: str) -> "Edition | None":  # type: ignore[return]
                     return web.ctx.site.get(matches[0])
 
         # Attempt to fetch the book from the import_item table
-        if result := ImportItem.import_first_staged(identifiers=isbns):
-            return result
+        if edition := ImportItem.import_first_staged(identifiers=isbns):
+            return edition
 
-        # TODO: Final step - call affiliate server, with retry code migrated there.
+        # Finally, try to fetch the book data from Amazon + import, using
+        # priority=0 to put the request at the front of the queue.
+        try:
+            r = requests.get(
+                f'http://{affiliate_server_url}/isbn/{isbn10 or isbn13}?priority=0'
+            )
+            r.raise_for_status()
+            if data := r.json().get('hit'):
+                import_item = {
+                    'ia_id': data['source_records'][0],
+                    'status': 'staged',
+                    'data': data,
+                }
```

Comment:
You were definitely right the first time!

`_get_amazon_metadata` is the right function, however you're also right that we do need a flag for e.g. `?priority` as an option both for (a) the affiliate server and (b) we probably want to use the `priority` option  instead of `sleep` + `retries` to have `_get_amazon_metadata` be able to use the `priority` mode (of waiting for the answer to be ready) rather than retrying. 

https://github.com/internetarchive/openlibrary/blob/a014cedcbfe8b06e5be076983ae5de9ff1ce9279/openlibrary/core/vendors.py#L306-L342

Reply:
I've consolidated around `get_amazon_metadata()` such that the calls to the affiliate server go through there now, and use `?high_priority=true`.

For the time being I have done away with all retries for `get_amazon_metadata()` / `_get_amazon_metadata()`. Instead, `_get_amazon_metadata()` calls the affiliate server, which:
- returns the item from cache if possible;
- if the item is not in cache, it's put into a queue (with `high_priority` items going to the front of the queue); and
- the `/isbn/` endpoint on the affiliate server will check the cache three times and if possible return the product data to `_get_amazon_metadata()`.

It doing this, I think, regardless of whether this PR is merged, calls to the affiliate server are completely blocking the server from sending the template back to the browser initiating the request, and I don't think making the requests async will help because the template can't be sent until it has this data.

*IF* this is the case, perhaps rather than have the backend put price data into the template, the back end could make an async request to get the product data in the affiliate server cache, and JavaScript on the client could request it from the cache later, as a final part of loading the page? If this is all correct, and we're making one request per second to the affiliate server, that's quite a lot of blocking before anything can happen, as the browser doesn't even have the template at this point.